api/controllers/AuthController.js

```diff
@@ -782,19 +865,21 @@ module.exports = {
             fail: true,
           },
         );
-        throw Boom.badRequest('Missing authorization code');
+        return Boom.unauthorized('Missing authorization code');
```

Comment:
I am converting any error where we know the reason to `return Boom...` instead of throwing. The throws were causing us to return 500s in tons of situations where 400/401 were the appropriate response.

Reply:
In testing this on my local GHO, I didn't see any ill effects. There's still an ugly PHP error when config is wrong, but maybe you can provide some Drupal insight on how to prevent it. Here's a sample drupal error I made by falsifying the `client_secret` and attempting OAuth login:

```
The website encountered an unexpected error. Please try again later.

League\OAuth2\Client\Provider\Exception\IdentityProviderException: Bad Request in League\OAuth2\Client\Provider\HumanitarianId->checkResponse() (line 124 of /srv/www/vendor/un-ocha/oauth2-hid/src/Provider/HumanitarianId.php).
League\OAuth2\Client\Provider\AbstractProvider->getParsedResponse(Object) (Line: 537)
League\OAuth2\Client\Provider\AbstractProvider->getAccessToken(Object, Array) (Line: 31)
Drupal\social_auth_hid\HidAuthManager->authenticate() (Line: 250)
Drupal\social_auth\Controller\OAuth2ControllerBase->processCallback() (Line: 70)
Drupal\social_auth_hid\Controller\HidAuthController->callback()
call_user_func_array(Array, Array) (Line: 123)
Drupal\Core\EventSubscriber\EarlyRenderingControllerWrapperSubscriber->Drupal\Core\EventSubscriber\{closure}() (Line: 573)
Drupal\Core\Render\Renderer->executeInRenderContext(Object, Object) (Line: 124)
Drupal\Core\EventSubscriber\EarlyRenderingControllerWrapperSubscriber->wrapControllerExecutionInRenderContext(Array, Array) (Line: 97)
Drupal\Core\EventSubscriber\EarlyRenderingControllerWrapperSubscriber->Drupal\Core\EventSubscriber\{closure}() (Line: 151)
Symfony\Component\HttpKernel\HttpKernel->handleRaw(Object, 1) (Line: 68)
Symfony\Component\HttpKernel\HttpKernel->handle(Object, 1, 1) (Line: 57)
Drupal\Core\StackMiddleware\Session->handle(Object, 1, 1) (Line: 47)
Drupal\Core\StackMiddleware\KernelPreHandle->handle(Object, 1, 1) (Line: 106)
Drupal\page_cache\StackMiddleware\PageCache->pass(Object, 1, 1) (Line: 85)
Drupal\page_cache\StackMiddleware\PageCache->handle(Object, 1, 1) (Line: 47)
Drupal\Core\StackMiddleware\ReverseProxyMiddleware->handle(Object, 1, 1) (Line: 52)
Drupal\Core\StackMiddleware\NegotiationMiddleware->handle(Object, 1, 1) (Line: 23)
Stack\StackedHttpKernel->handle(Object, 1, 1) (Line: 708)
Drupal\Core\DrupalKernel->handle(Object) (Line: 19)
```

Reply:
@rupl The `League\OAuth2` library throws exceptions when something is wrong but the drupal `social_api`, `social_auth` and `social_auth_hid` modules don't seem to try to catch them.

After a little debugging, the error sent by the HID API looks like: 400 Bad Request with the a json payload containing the `message`: "invalid client_secret". 

However according to [some docs](https://www.oauth.com/oauth2-servers/server-side-apps/possible-errors/), it looks like the payload should have a `error_description` property instead of (or in addition to) the `message` one. 

This is also what the `un-ocha/oauth2-hid` is expecting:

```php
    protected function checkResponse(ResponseInterface $response, $data)
    {
        if (isset($data['error'])) {
            throw new IdentityProviderException(
                $data['error_description'] ?: $response->getReasonPhrase(),
                $response->getStatusCode(),
                $response
            );
        }
    }
  ```
  
 I would propose to change that if possible. That would not change the fact that the `social_*` modules don't handle the exceptions but at least might be closer to the specs.
 
 As for preventing the PHP error, I think we could add some exception catching to the `social_auth_hid` module and redirect to the homepage or something like that and log the error message? cc @attiks

Reply:
oh hey that's great feedback! In the `hapi-oauth2orize` plugin, there's a function `transformBoomError()` which appends that `error_description` field automatically. It sounds like I should be processing all our errors in this function through that transformer, in order to produce OAuth-compatible errors. I'll give it a shot.

Reply:
(just writing some notes to myself since this is getting de-prioritized)

In addition to the error format, the [OAuth possible errors](https://www.oauth.com/oauth2-servers/server-side-apps/possible-errors/) doc from Jonathan mentions some redirects when they are both possible and safe. Can't hurt to implement those too.